From fea0a2cc1f3d48ae6eb6dcb95e439b13a657580d Mon Sep 17 00:00:00 2001
From: Bharat Bhushan <bbhushan2@marvell.com>
Date: Thu, 31 Oct 2019 13:52:23 +0530
Subject: [PATCH 0668/1921] coresight: Add missing unregister of non-secure
 buffer

Non-Secure buffer registration will fail on next run if current
non-secure buffer overlap with previouly registered non-secure
buffer. That will lead to ATF crash as it will access un-mapped
non-secure buffer.

Change-Id: I3caa8b62b3b56ed8121c03c141f8478391e7576b
Signed-off-by: Bharat Bhushan <bbhushan2@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/32243
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Linu Cherian <lcherian@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/hwtracing/coresight/coresight-tmc-etr.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/hwtracing/coresight/coresight-tmc-etr.c b/drivers/hwtracing/coresight/coresight-tmc-etr.c
index a1f0d5776ac4..f5265c687e52 100644
--- a/drivers/hwtracing/coresight/coresight-tmc-etr.c
+++ b/drivers/hwtracing/coresight/coresight-tmc-etr.c
@@ -1478,6 +1478,9 @@ static void tmc_free_etr_buffer(void *config)
 	if (buf && WARN_ON(buf != etr_buf))
 		goto free_etr_perf_buffer;
 
+	tmc_unregister_drvbuf(drvdata, etr_buf->hwaddr,
+					      etr_buf->size);
+
 	tmc_free_etr_buf(etr_perf->etr_buf);
 
 free_etr_perf_buffer:
-- 
2.31.1

